src/Swarm/Game/Step.hs

```diff
@@ -419,7 +419,10 @@ entityAt loc = zoomWorld (W.lookupEntityM @Int (W.locToCoords loc))
 -- | Modify the entity (if any) at a given location.
 updateEntityAt ::
   (Has (State GameState) sig m) => Location -> (Maybe Entity -> Maybe Entity) -> m ()
-updateEntityAt loc upd = zoomWorld (W.updateM @Int (W.locToCoords loc) upd)
+updateEntityAt loc upd = do
+  didChange <- zoomWorld $ W.updateM @Int (W.locToCoords loc) upd
+  when didChange $
```

Comment:
Instead of using `when` here, I could pass `didChange` as an argument to `wakeWatchingRobots`, and lazy consumption could avoid performing expensive entity equality checks when a cell is unwatched.